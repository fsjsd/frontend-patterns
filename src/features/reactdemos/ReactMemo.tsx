/* eslint-disable react/display-name */
import React, { useState } from 'react'
import { useMemo } from 'react';


// eslint-disable-next-line react/display-name
const CounterAMemo: React.VFC<{ a: number }> = React.memo(({ a }) => {
  // console.count("CounterAMemo")
  // console.log("CounterAMemo", { a })
  return (
    <div>
      <p>CounterAMemo a={a}</p>
    </div>
  )
})

const CounterABMemo: React.VFC<{ myObj: { a: number, b: number } }> = React.memo(({ myObj }) => {
  // console.count("CounterABMemo")
  // console.log("CounterABMemo", { ...myObj });
  return (
    <div>
      <p>CounterABMemo a={myObj.a} b={myObj.b}</p>
    </div>
  )
});

const CounterABUseMemo: React.VFC<{ myObj: { a: number, b: number } }> = ({ myObj }) => {
  // this is neat but still executes js in component
  return useMemo(() => {
    // console.count("CounterABUseMemo")
    // console.log("CounterABUseMemo", { ...myObj })
    return (
      <div>
        <p>CounterABUseMemo a={myObj.a} b={myObj.b}</p>
      </div>
    )
  }, [myObj.a, myObj.b]);
}

const deepMemoComparator = <T,>(prev: Readonly<React.PropsWithChildren<T>>, next: Readonly<React.PropsWithChildren<T>>): boolean => {
  // const propKeys = Object.keys(prev).filter(key => key !== "children");
  const areEqual: boolean = Object.keys(prev).every((key) => {
    if (key === "children") {
      return true; // skip React children prop
    } else if (typeof prev[key as keyof T] === "object") {
      return deepMemoComparator(prev[key as keyof T], next[key as keyof T])
    } else {
      return prev[key as keyof T] === next[key as keyof T]
    }
  });
  //console.log(Object.keys(prev), prev, next, areEqual)
  return areEqual;
}

/*
const shallowMemoComparator = <T,>(prev: Readonly<React.PropsWithChildren<T>>, next: Readonly<React.PropsWithChildren<T>>) => {
  const areEqual = Object.keys(prev).filter(key => key !== "children").every((key) => prev[key as keyof T] === next[key as keyof T]);
  // console.log(Object.keys(prev), prev, next, areEqual)
  return areEqual;
}
*/

const CounterABShallowMemo: React.FC<{ myObj: { a: number, b: number } }> = React.memo(({ myObj, children }) => {
  // console.count("CounterABShallowMemo")
  // console.log("CounterABShallowMemo", { ...myObj });
  return (
    <div>
      <p>CounterABShallowMemo a={myObj.a} b={myObj.b}</p>
      {children}
    </div>
  )
}, deepMemoComparator);



const ReactMemo = () => {
  const [countObj, setCountObj] = useState({
    a: 0, b: 0, c: () => {
      // no op
    }
  });

  const handleIncAClick = () => {
    setCountObj((prev) => ({ ...prev, a: prev.a + 1 }))
  };
  const handleIncBClick = () => {
    setCountObj((prev) => ({ ...prev, b: prev.b + 1 }))
  };
  const handleNoOpClick = () => {
    setCountObj((prev) => ({ ...prev }))
  };


  return (
    <div role="main" style={{ padding: "10px" }}>
      <CounterAMemo a={countObj.a} />
      {/* immutable objects re-created through react state will cause full re-render */}
      <CounterABMemo myObj={countObj} />
      {/* but memo comparing specific properties won't re-render if no change */}
      <CounterABUseMemo myObj={countObj} />
      {/* deep memo compares specific object keys */}
      <CounterABShallowMemo myObj={countObj}>children!</CounterABShallowMemo>

      <p>Container state: {countObj.a}</p>
      <button className='bg-gray' onClick={() => handleIncAClick()}>Inc A</button>
      <button className='bg-gray' onClick={() => handleIncBClick()}>Inc B</button>
      <button className='bg-gray' onClick={() => handleNoOpClick()}>Null state change</button>
    </div>
  )
}

export default ReactMemo
