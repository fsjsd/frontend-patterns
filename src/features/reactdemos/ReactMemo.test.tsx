import React from "react";
import ReactMemo, { CounterABDeepMemo, CounterABMemo, CounterABUseMemo, CounterAMemo } from "./ReactMemo";
import { fireEvent, render, waitFor } from "@testing-library/react";

describe("ReactMemo", () => {
  it("renders and matches snapshot", async () => {
    const { container, queryByRole } = render(<ReactMemo />);
    await waitFor(() => expect(queryByRole("main")).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });

  it("changes state after increments", async () => {
    const { container, queryByRole, getByRole } = render(<ReactMemo />);
    await waitFor(() => expect(queryByRole("main")).toBeInTheDocument());
    expect(container).toMatchSnapshot();
    fireEvent.click(getByRole("button", { name: "Inc A" }));
    expect(container).toMatchSnapshot();
    fireEvent.click(getByRole("button", { name: "Inc B" }));
    expect(container).toMatchSnapshot();
    fireEvent.click(getByRole("button", { name: "Null state change" }));
    expect(container).toMatchSnapshot();
  });

  it("memo component does not re-render if props don't change", async () => {
    const props = { a: 1, b: 1 };
    const renderCount = jest.fn();
    const { rerender } = render(<CounterAMemo a={props.a} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(1);
    // re-assign props with same primitive value for 'a'
    props.a = 1;
    props.b = 2;
    rerender(<CounterAMemo a={props.a} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(1);
  });

  it("memo component re-renders if object prop changes (with same primitive values)", async () => {
    const props = { myObj: { a: 1, b: 1 } };
    const renderCount = jest.fn();
    const { rerender } = render(
      <CounterABMemo myObj={props.myObj} render={renderCount} />
    );
    expect(renderCount).toBeCalledTimes(1);
    rerender(<CounterABMemo myObj={props.myObj} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(1);

    // assign completely new object with same values
    props.myObj = { a: 1, b: 1 };
    rerender(<CounterABMemo myObj={props.myObj} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(2);
  });

  it("usememo component with custom comparator does not re-renders if object prop changes (with same primitive values)", async () => {
    const props = { myObj: { a: 1, b: 1 } };
    const renderCount = jest.fn();
    const { rerender } = render(
      <CounterABUseMemo myObj={props.myObj} render={renderCount} />
    );
    expect(renderCount).toBeCalledTimes(1);
    rerender(<CounterABUseMemo myObj={props.myObj} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(1);

    // assign completely new object with same values
    props.myObj = { a: 1, b: 1 };
    rerender(<CounterABUseMemo myObj={props.myObj} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(1);
  });

  it("memo component with deep comparator does not re-renders if object prop changes (with same primitive values)", async () => {
    const props = { myObj: { a: 1, b: 1 } };
    const renderCount = jest.fn();
    const { rerender } = render(
      <CounterABDeepMemo myObj={props.myObj} render={renderCount} />
    );
    expect(renderCount).toBeCalledTimes(1);
    rerender(<CounterABDeepMemo myObj={props.myObj} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(1);

    // assign completely new object with same values
    props.myObj = { a: 1, b: 1 };
    rerender(<CounterABDeepMemo myObj={props.myObj} render={renderCount} />);
    expect(renderCount).toBeCalledTimes(1);
  });




});
