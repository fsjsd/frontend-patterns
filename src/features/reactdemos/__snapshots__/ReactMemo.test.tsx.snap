// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactMemo changes state after increments 1`] = `
<div>
  <div
    role="main"
    style="padding: 10px;"
  >
    <div>
      <p>
        CounterAMemo a=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABMemo a=
        0
         b=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABUseMemo a=
        0
         b=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABShallowMemo a=
        0
         b=
        0
      </p>
      children!
    </div>
    <p>
      Container state: 
      0
    </p>
    <div
      class="sc-gsnTZi jwMZeR"
    >
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc A
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc B
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Null state change
      </button>
    </div>
  </div>
</div>
`;

exports[`ReactMemo changes state after increments 2`] = `
<div>
  <div
    role="main"
    style="padding: 10px;"
  >
    <div>
      <p>
        CounterAMemo a=
        1
      </p>
    </div>
    <div>
      <p>
        CounterABMemo a=
        1
         b=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABUseMemo a=
        1
         b=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABShallowMemo a=
        1
         b=
        0
      </p>
      children!
    </div>
    <p>
      Container state: 
      1
    </p>
    <div
      class="sc-gsnTZi jwMZeR"
    >
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc A
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc B
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Null state change
      </button>
    </div>
  </div>
</div>
`;

exports[`ReactMemo changes state after increments 3`] = `
<div>
  <div
    role="main"
    style="padding: 10px;"
  >
    <div>
      <p>
        CounterAMemo a=
        1
      </p>
    </div>
    <div>
      <p>
        CounterABMemo a=
        1
         b=
        1
      </p>
    </div>
    <div>
      <p>
        CounterABUseMemo a=
        1
         b=
        1
      </p>
    </div>
    <div>
      <p>
        CounterABShallowMemo a=
        1
         b=
        1
      </p>
      children!
    </div>
    <p>
      Container state: 
      1
    </p>
    <div
      class="sc-gsnTZi jwMZeR"
    >
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc A
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc B
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Null state change
      </button>
    </div>
  </div>
</div>
`;

exports[`ReactMemo changes state after increments 4`] = `
<div>
  <div
    role="main"
    style="padding: 10px;"
  >
    <div>
      <p>
        CounterAMemo a=
        1
      </p>
    </div>
    <div>
      <p>
        CounterABMemo a=
        1
         b=
        1
      </p>
    </div>
    <div>
      <p>
        CounterABUseMemo a=
        1
         b=
        1
      </p>
    </div>
    <div>
      <p>
        CounterABShallowMemo a=
        1
         b=
        1
      </p>
      children!
    </div>
    <p>
      Container state: 
      1
    </p>
    <div
      class="sc-gsnTZi jwMZeR"
    >
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc A
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc B
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Null state change
      </button>
    </div>
  </div>
</div>
`;

exports[`ReactMemo renders and matches snapshot 1`] = `
<div>
  <div
    role="main"
    style="padding: 10px;"
  >
    <div>
      <p>
        CounterAMemo a=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABMemo a=
        0
         b=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABUseMemo a=
        0
         b=
        0
      </p>
    </div>
    <div>
      <p>
        CounterABShallowMemo a=
        0
         b=
        0
      </p>
      children!
    </div>
    <p>
      Container state: 
      0
    </p>
    <div
      class="sc-gsnTZi jwMZeR"
    >
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc A
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Inc B
      </button>
      <button
        class="sc-bczRLJ daJtMB bg-gray"
      >
        Null state change
      </button>
    </div>
  </div>
</div>
`;
